Add Sigsum support to tkey-verification

mchack-work opened the pull request:
## Description

Main goal: Add Sigsum support to tkey-verification. See docs/design.md for our thoughts on how we do this.

- Add Sigsum transparency log support. See https://sigsum.org/
- Remove vendor signing with vendor keys and instead uses the Sigsum
  key pair. Vendor key support is retained for verification.
- Introduce lifetime of vendor keys, but limited to Sigsum submit
  keys.
- Split out verification by end user to its own program: tkey-verify.
- Introduce new program: tkey-sigsum-submit.
- Lots of refactoring, again.
- Include new signer app, signer-tkey-verification1, built from
  https://github.com/tillitis/tkey-device-signer without touch. This
  version supports both the older Bellatrix and the upcoming Castor
  TKeys.

Fixes #31

I think a review should focus on:

- Getting the entire flow up and running:
  1. tkey-verification remote-sign speaking to serve-sign
  2. Submitting to the Sigsum log with tkey-sigsum-submit.
  3. Verify with tkey-verify.

- Understand the design (see doc/design.md).
- Compare the code with the design.

There are just too numerous refactoring things that I think a review should focus on the above, and not necessarily all the details of the implementation.

## Type of change

- [x] Breaking Change (a change which would cause existing functionality to not work as expected)
- [x] Documentation (a change to documentation)

## Submission checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my changes
- [x] I have tested and verified my changes on target
- [x] My changes are well written and CI is passing
- [x] I have squashed my work to relevant commits and rebased on main for linear history
- [x] I have added a "Co-authored-by: x" if several people contributed, either pair programming or by squashing commits from different authors.
- [x] I have updated the documentation where relevant (readme, dev.tillitis.se etc.)
- [ ] QEMU is updated to reflect changes

Internal documentation is updated, not yet the website introduction, but that can wait until a release.

mchack-work committed:
Revise README: Describe how it will be used with Sigsum

mchack-work committed:
Add experimental sigsum support to verify

mchack-work committed:
Support both vendor signatures and Sigsum proofs

mchack-work committed:
Add test data for sigsum verification

mchack-work committed:
doc: Escape newlines in JSON

agren committed:
Add tool to sigsum log devices and generate verification files

Takes a sigsum submit request from each file in the `submissions-dir`
directory and submits it to the log. Each submission file generates a
corresponding verification file. The verification files are written to
the `verifications-dir` directory.

mchack-work committed:
Simplify errors

mchack-work committed:
Move Verification type to own package

- Let the type live in its own package
- Add some parsing functions on the type.

agren committed:
Test sigsum log submission

mchack-work committed:
Re-arrange to internal packages

Move:

- appbins
- vendorpubkeys.go -> vendorkey

to their own packages.

Rewrite package verification to use the above, to be more picky in
parsing to native types and support both types of verifications.

mchack-work committed:
Add VerifyProofDigest() for verifiying the raw hash

mchack-work committed:
Add initial tests for verification

- Parsing test
- Proof verification with digest only

agren committed:
Update signing procedure diagram use Sigsum

agren committed:
log-submit: Switch to using the new verification package

* Use the new helper methods on Verification (Which has moved to the new
verification package
* Add and use Verification.toJSON/.ToFile
* Move the Submission type to it's own package and change it to match
the structure of the new Verification and VerificatioJson types

mchack-work committed:
doc: Revise README

Describe the new system in both a better overview and more detail.

Include open questions.

agren committed:
serve-signer: Create submission request file instead of verification file

Instead of outputting a file with an ed25519 signature we output a file
with a Sigsum submit request

mchack-work committed:
doc: Revise README

- Change "provider" back to "vendor".
- Change control the identity to check identity.

mchack-work committed:
vendorkey: Prepare to be initialized from different sources

- Add a FromEmbedded() function to get the embedded vendor keys.

- Add a FromString() function to use for testing our from dynamic
  configuration.

mchack-work committed:
Add simple test to verify vendor signatures

mchack-work committed:
Add missing arg to fmt.Errorf()

mchack-work committed:
Add test to verify message with proof, not just raw hash

mchack-work committed:
Lose debug print

mchack-work committed:
Clean up verify()

Make it easier to follow and understand what it does.

Move challenge/response to the tkey package.

agren committed:
remotesign: Use new tk.Challenge method

agren committed:
Assume latest Bellatrix firmware when detecting a default TKey

Assume that TKeys with the default UDI 0001020304050607 have the latest
firmware (TK1-24.03).

agren committed:
Add state to fake sigsum server

agren committed:
Test Submission JSON encoding/decoding

agren committed:
verify: Verify sigsum proof using default TKey pubkey for now

Use the pubkey of a TKey built using the default UDS while developing.

agren committed:
submit: Move submission, verification and policy test strings to files

Submission file built from two TKeys with default UDS. One used as
device to submit to the log, and the other used as vendor signing key.

Tkey-verification was run like so:
- go run ./cmd/tkey-verification remote-sign --port /dev/tty-tkey-b-test-device --config tkey-verification.yaml.example-remote-sign
- go run ./cmd/tkey-verification/ serve-signer --port /dev/tty-tkey-b-test-vendor-key --config tkey-verification.yaml.example-serve-signer

agren committed:
submit: Extract file processing loop from main()

agren committed:
submit: Remove proof debug print

mchack-work committed:
Update tkeyclient and its deps

mchack-work committed:
verify: Introduce --sigsum to demand a Sigsum proof

Always the default when vendor is Tillitis with product ID == Castor.

agren committed:
submit: Add parameterized test

agren committed:
submit: Move submission files once processed

Each submission file is moved when the resulting verification file has
been created

agren committed:
submit: Abort if verification directory is not empty

agren committed:
submit: Name test directories for easier debugging

agren committed:
submit: Abort if processed submissions directory is not empty

agren committed:
submit: Move test parameters to named struct for readability

agren committed:
submit: Make sure all submission files can be opened before logging anything

mchack-work committed:
docs: Revise README

- Describe why we want to use a Sigsum log.

- Add at least something about how we plan to use a Sigsum monitor.

- Fix details about what we sign.

- Fix details in structure.

- Remove something that looked like HTML but was really a
  metasyntactic variable.

agren committed:
submit: Move all test samples to the same directory

agren committed:
submit: Break out test sample copying code

agren committed:
Add missing err message

agren committed:
submit: Add tkey-sigsum-submit to Makefile

mchack-work committed:
Move doc string to correct function

mchack-work committed:
Introduce SigsumLog type and methods

- New type SigsumLog to describe a Sigsum transparency logs, its keys
  and and policies.

- Use new type and methods in verify command and verification tests.

agren committed:
submit: Move repetitive test assertions into helper functions

- Adds assertDirContainsOnly, assertErrorMsgStartsWith
- requireFileContentsEqual changes name to assertFileContentsEqual an no
  longer stops the test early when files are not equal. Reason being
  that other assertions might give useful output.

agren committed:
submit: Mark assertions as test helpers to get useful line numbers on failures

mchack-work committed:
Move Sigsum static data to internal/data

mchack-work committed:
Move firmwares.go to package internal/firmware

- Add function FromJson()
- Add static data with firmwares in internal/data/data.go
- Add simple parsing tests to firmware package.
- Add firmware embedded data to internal/data

mchack-work committed:
Refactor remoteSign()

- Rename signChallenge() -> authDevice(). Use a message struct for
  constructing the message.

- Do the device authentication before verifying firmware hash.

- Require that firmwares is filled in before calling
  verifyFirmwareHash() with it as a new argument instead of filling it
  inside the function.

mchack-work committed:
Add SPDX to verification

mchack-work committed:
Break out tkey-verify as its own program

mchack-work committed:
Move embedded vendor pubkey to internal/data

mchack-work committed:
doc: Update location of embedded device apps

mchack-work committed:
Change types for appbin hashes

- Use byte array instead of slice for SHA512 digest which identifies a
  specific device app.

- Remove Get() from appbins, access the map directly with the digest,
  which is now possible since it's no longer a slice.

- Introduce a util package with decoding function for hashes from
  string.

mchack-work committed:
firmwares: Store embedded firmware data as simple lines instead

- Introduce FromString() and MustDecodeString() to package firmware.

- Change storage of firmware data in internal/data to remind us of how
  it used to look to minimize problems converting to JSON.

- Add test for firmware.FromString().

- Update Maintenance in README.

mchack-work committed:
submit: Use embedded policy from internal/data

mchack-work committed:
Add test Sigsum submit key

mchack-work committed:
Teach show-pubkey how to print the pubkey in SSH format

Borrow some BSD2 code from:

https://git.glasklar.is/sigsum/core/sigsum-go

for the new internal/ssh package to be able to do this.

agren committed:
submit: Document test

agren committed:
doc: Add tkey-sigsum-submit to Testing chapter

agren committed:
doc: Clarify what data is sent to the monitor

agren committed:
doc: Replace inline example files with links to test files

agren committed:
doc: Document submit request file

mchack-work committed:
submit: Update help text

Change -n description from earlier copy/paste.

mchack-work committed:
Add man page for tkey-sigsum-submit

mchack-work committed:
doc: Update tkey-verify man page

- Fix header showing wrong command
- Simplify language
- List options as own section
- Fix markup

mchack-work committed:
doc: Update man page for tkey-verification

- Remove verify command.
- Simplify language.
- Describe the new submission file format.
- Add missing --app argument to show-pubkey.
- Remove output from show-pubkey in FILES.
- Add SEE ALSO.

agren committed:
submit: Add spdx to .go files

agren committed:
spdx: Change types for appbin hashes - util.go

agren committed:
submit: Add testdata to REUSE.toml

agren committed:
spdx: Add man page for tkey-sigsum-submit

agren committed:
spdx: Re-arrange to internal packages - Move bins to internal/appbins

agren committed:
spdx: doc/tkey-verify.*

agren committed:
spdx: internal/ssh/README.md

agren committed:
spdx: Remove stray verifications folder

agren committed:
Remove spdx-ensure since its functionality is covered by REUSE.toml

agren committed:
gitignore: tkey-verify

agren committed:
gitignore: tkey-sigsum-submit

agren committed:
Fix out-of-sync comments in firmware_test:TestWrongFirmware

agren committed:
verify: Fill in vendorkeys & Sigsum config early

- Instead of filling in configuration in functions under verify(),
  fill them in early and exit early if something's wrong.

- Remove now unnecessary verifyProof() and verifySig() functions.

Co-authored-by: Michael Cardell Widerkrantz <mc@tillitis.se>

agren committed:
Move separate buildMessage implementations to util

Moved from:

- tkey-verification/common.go
- tkey-verify/verify.go
- internal/verification_test.go

to util/util.go

agren committed:
Tidy tkey-verify.scd

mchack-work committed:
Add --version to tkey-verify & tkey-sigsum-submit

- Introduce util.Version() to get the version either set from Makefile or
  from Go module.

- Use same function in tkey-verification.

mchack-work committed:
Update golangci-lint config to new config format

mchack-work committed:
ci: Update to new golangci-lint

mchack-work committed:
ci: Remove SPDX check

tools/spdx-ensure has been removed. We rely on REUSE compliance test now.

agren committed:
Follow initialism and acronym style rules

agren committed:
Follow perfsprint linting rules

Changing:

- Only use fmt.Errorf when formatting is needed

- Concatenate strings instead of havining %s at end of string

agren committed:
Follow errorlint linting rules

- Use lower case error messages

- Use %w when wrapping errors with fmt.Errorf

agren committed:
Remove unused functions

agren committed:
Fix typos

agren committed:
Check error

agren committed:
Follow revive indent-error-flow linting rule

agren committed:
Use bytes.Equal instead of bytes.Compare unless necessary

agren committed:
Create files with 0600 mode instead of 0644

agren committed:
Remove todo

agren committed:
Silence gosec false positives

Gosec linter complains about unchecked overflow when type conversion,
from Uint32 to Uint8 and Uint16, when decoding the UDI:

- "G115: integer overflow conversion uint32 -> uint8"
- "G115: integer overflow conversion uint32 -> uint16"

The type conversion will not fail since the values are truncated by the
and operation. Silencing the false positive.

agren committed:
Disable testpackage linter

agren committed:
Follow wrapcheck linting rules

Wrapping external errors with fmt.Errorf

agren committed:
Silence gosec false positive

Gosec complains about "G115: integer overflow conversion int -> uint32".

Instead of calling len(s) multiple times. Let's call len(s) once and
check the result for overflow before using it.

agren committed:
lint: Allow package name "util"

agren committed:
Follow revive linting rules

Avoid "exported: ... stuttering" errors by renaming:

- sigsum.SigsumLog to sigsum.Log
- verification.verificationType to verification.Type

agren committed:
Follow govet shadow rule

Removes shadowing of err variables.

mchack-work committed:
Change Sigsum submit key configuration

Add support for many submit keys, configure the active one with
activekey in config. After loading the associated device app verify
that the received public key is the expected one.

For configuration we use a constant string with a simple format like
this:

  name
  ssh key
  tag of app
  app SHA512

We repeat this for all the Sigsum submit keys we allow for
verification.

mchack-work committed:
Add support for Sigsum submit key lifetimes

Fail verification if a Sigsum proof contains witness signatures
outside of the lifetime of the used submit key.

Co-authored-by: Mikael Ågren <mikael@tillitis.se>

agren committed:
Increase test submit key lifetime to ~100 years

agren committed:
submit: Verify witness timestamps

The sigsum library already checks that the proof received when
submitting is valid. We add a call to VerifyProofDigest to also check
the timestamps reported by the witnesses are withing the lifetime of the
submit key.

agren committed:
submit: Use policy from embedded sigsum configuration

agren committed:
verify: Allow forcing sigsum verification using flag --sigsum

agren committed:
Return error if submit key is not found

agren committed:
Print sigsum submit key used in verified proof

Changing VerifyProofDigest() to return the used submit key

mchack-work committed:
Change type of argument ssh public key helper functions

Use a simple [ed25519.Size]byte for public keys

mchack-work committed:
Use own copy of Sigsum log config in submission tests

Refactor submission functions to be functions with a receiver type be
able to do this easier.

agren committed:
doc: Update vendor's public key chapter to mention sigsum

- Use SigsumConf instead of VendorPubKeys
- App binaries have moved to internal/appbins/bins
- Update show-pubkey example command
- Use SSH format instead of hex format

mchack-work committed:
Add new signer with tag tkey-verification1

- signer built with TKEY_SIGNER_APP_NO_TOUCH=yes from

  https://github.com/tillitis/tkey-device-signer/

- Using tag tkey-verification1

- Set as active for remote-sign.

mchack-work committed:
doc: Update tkey-verification man page

Describe the configuration files to both serve-signer and remote-sign.

mchack-work committed:
doc: Update README about maintainance

- Mention the Sigsum configuration under Maintenance.

- Describe how to add a new device app, both for remote-sign and, as a
  new key for serve-signer.

mchack-work committed:
build: Bump to tkey-builder:5rc2

mchack-work committed:
doc: Revise README, break out design, integrate implementation notes

- Remove a lot of old vendor key stuff and mistakes like mentioning
  tkey-verification when it should have been tkey-verify.

- Provide different usage instructions for end user and vendor.

- Update provisioning instructions.

- Simplify the test setup description.

- Move the TKey verification process design to its own file. Link to
  it from README.

- Integrate some of the implementation notes under maintenance and
  remove the old doc/implementation-notes.md.

agren committed:
Reenable --verbose flag for remote-sign

mchack-work committed:
ci: Add go test

agren committed:
doc: fixup

mchack-work committed:
Remove -build flag from tkey-verify

End user, even advanced end users, might not need to know exactly what
assets are built in to tkey-verify. If an advanced user really want to
know it should be possibly to find with --version and then checking in
the code.

mchack-work committed:
Add Sigsum config to --build

We introduce a .String() function on the Log type to help us do this.

mchack-work committed:
doc: Add release notes for upcoming release

agren committed:
doc: Fix links

agren committed:
doc: Remove duplicate monitor step

Removing the last step and keeping step 8 since the monitor need to know
about the produced device before it is logged.

dehanj committed:
Remove flags that are remainders from the split to tkey-verify

dehanj committed:
submit: Make sure post-submission and verification dir are different

Otherwise the verification file will be overwritten by the submission
file.

dehanj committed:
submit: make the help message a bit cleaner

dehanj committed:
doc: fix typos

dehanj reviewed (approved):
I have done my review, including:
Getting the entire flow up and running:
- tkey-verification remote-sign speaking to serve-sign
- Submitting to the Sigsum log with tkey-sigsum-submit.
- Verify with tkey-verify.

- Gone through the documentation and left a few comments
- I have made some commits, mostly typos and some fixes/clean up. Nothing major. Please look them through before any merging. 

I have not gone through every piece of code. 

I have a few questions, but they are more tied to Sigsum than tkey-verification so will ask them in another forum.

mchack-work committed:
doc: Revise design document

- tkey-verify does the verification.

- Remove text about the checksum at the monitor. It is already covered
  by point 8.

- Avoid unnecessary explanation of what the verification file
  contains.
